### ISSUE TITLE
EF Core maps DateTime type to TEXT in SQLite

### ISSUE THREAD

**Issue by sgurukrupa** [aspnet/EntityFramework.Docs#3080]
This may be considered one more limitation, but it may not. Nevertheless, it's good to document that the provider available in the Nuget package Microsoft.EntityFrameworkCore.SQLite, Version 5.0.1 maps C# DateTime objects to TEXT in SQLite. How this design caught me off my feet is documented [here](https://sgurukrupa.wordpress.com/2021/02/04/ef-core-maps-c-datetime-type-to-text-in-sqlite/).

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: dbc4d900-28a8-9c04-2443-7407f15a684f
* Version Independent ID: 545bb0b5-f15b-584e-cef7-7283a0e0729f
* Content: [SQLite Database Provider - Limitations - EF Core](https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations)
* Content Source: [entity-framework/core/providers/sqlite/limitations.md](https://github.com/dotnet/EntityFramework.Docs/blob/master/entity-framework/core/providers/sqlite/limitations.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @bricelam
* Microsoft Alias: **avickers**

**Comment by AndriySvyryd** [comment]
This is a limitation of SQLite, not EF Core. What column type did you expect?

**Comment by sgurukrupa** [comment]
I expected that C# DateTime would be mapped by EF Core to some type that had a notion of ordering in terms of time, for example, an integer that represents say the number of ticks since 1970-01-01 00:00:00 UTC. And, yes, I agree with you that this is more a limitation of SQLite, rather than EF Core. But, it wouldn't harm anyone to explicitly document that EF Core maps C# DateTime to TEXT in SQLite, either. It took me some time to dig out this fact from the Internet.

**Comment by bricelam** [comment]
> some type that had a notion of ordering in terms of time

Note that the TEXT format works with order and comparison operators. Adding a [value converter](https://docs.microsoft.com/ef/core/modeling/value-conversions) to convert all values UTC would also make the operators work when you have a mix of DateTimeKind.Local and Utc on the client.

**Comment by sgurukrupa** [comment]
> Adding a value converter to convert all values UTC would also make the operators work when you have a mix of DateTimeKind.Local and Utc on the client.
But, I won't be able to make this work on the [database server side](https://docs.microsoft.com/en-us/ef/core/querying/client-eval), would I? All this computation would have to happen on the client side, for example, if want to compute the [max](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.max?view=net-5.0) on a column of dates.

**Comment by bricelam** [comment]
The idea is to normalize all DateTime values in the database to UTC so you can perform operations on the server.

**Comment by janseris** [comment]
Is there any update on this? Having DateTime scaffolded as string isnt really cool for an ORM.

**Comment by ErikEJ** [comment]
@janseris What SQLite "data type" would you prefer?

**Comment by janseris** [comment]
@ErikEJ 

`entity.Property(e => e.ValidToUtc).HasColumnType("DATETIME");`

such column being exposed as (UTC) `DateTime`/`DateTime?` for the C# code (stored as string sadly I have to use SQLite and not MSSQL).
Right now, it is generated on C# side as `string` using EF Core Power Tools 2.6.1604 (for DB -> C# code generating) and EF Core Power Pack 2.6.1603 (for SQLite support).

The SQLite column definition is:
`ValidToUtc DATETIME NULL,` which from my very small experience with SQLite is probably TEXT with some automatic metadata/constriants?

**Comment by ErikEJ** [comment]
There is no "DATETIME" data type in SQLite! https://www.sqlite.org/datatype3.html

**Comment by janseris** [comment]
Somehow it seems it is

<img width="687" height="478" alt="Image" src="https://github.com/user-attachments/assets/2dcc9d8b-1aa1-42b0-ba99-39c0cf60c644" />

**Comment by ErikEJ** [comment]
@janseris Only in tools and drivers - not in the base engine

**Comment by janseris** [comment]
However it seems like (I am sorry I discovered it now) that it is only a made-up value which can be anything. 
I believe then TEXT is the real one behind the scenes which I need to expose to C# as DateTime from SQLite any nice/standard way possible

This issue seems to be the C# DateTime -> TEXT way and mine seems to be TEXT -> C# DateTime using EF Core Power Tools, maybe we should be solving it in your repository issues, @ErikEJ?
I looked now into Advanced in settings in EF Core Power Tools but there doesn't seem to be an option for this.
There doesn't seem to be a limitation for `DateTime` data type listed in https://learn.microsoft.com/en-us/ef/core/providers/sqlite/limitations

**Comment by ErikEJ** [comment]
@janseris Yes, feel free to create a feature request in the EF Core Power Tools repo

**Comment by janseris** [comment]
Created a feature request in EF Core Power Tools repo which contains a working sample how to achieve it now using EF Core Power Tools for anyone trying to find out how to do it.
https://github.com/ErikEJ/EFCorePowerTools/issues/3470
